### ISSUE TITLE
Missing node ?

### ISSUE THREAD

**Issue by bohtho** [gorenje/node-red-contrib-nodedev#1]
Is there a missing node (and instructions) for the Github functionality ? A designated end node in parallell to the NR Install and NPM publish nodes ?

The only way I could make it work was by linking a NodeFactory generated node dev flow to the last half of the flow that commits to Github in the [Flowhub flow](https://flowhub.org/f/b92be5062203ff69) you refer to here. Only by looking through the code did I found out about the GITHUB_TOKEN (actually I later found it is informed about beneath that flow on Flowhub but not in the node readme).

And, btw, in that Flowhub flow I kept getting an error that I was missing a github token in my ENVs until I changed `process.env` to `get.env` in all occurences, but I guess that flow is "dated"(?) or not meant for the general public(?).

Also a thought: Since the NodeFactory has little function in the flow other than being a holder for the config fields and a button, couldn't that get implemented in the sidebar (plugin) too ? And split out the tgz converter if necessary ?

**Comment by gorenje** [comment]
> Is there a missing node (and instructions) for the Github functionality ?

Yes there is, it's this [flow](https://flowhub.org/f/d0506e991d512ace) that provides GitHub functionality. The [flow](https://flowhub.org/f/b92be5062203ff69) you're referencing duplicates that functionality because that flow *defines* the NodeDev nodes - so they can't use themselves, i.e. the [flow](https://flowhub.org/f/d0506e991d512ace) that provides the github functionality also uses the NodeDev nodes so it can't be used by a flow that defines the NodeDev nodes - makes perfect sense! 🤔 

And yes the documentation around GITHUB_TOKEN is admittedly thin ... 

> Since the NodeFactory has little function

The NodeFactory is designed to be an artifact in the flow that defines nodes. For example, this [flow](https://flowhub.org/f/bd2901f55cfc55ef) that defines the [FlowCompare](https://flows.nodered.org/node/@gregoriusrippenstein/node-red-contrib-flowcompare) node package contains the NodeFactory node as an indication of how the original node was generated (i.e. what properties). This is of course purely documentation however it did generated the template code for the node - in the beginning.

Also that [FlowCompare flow](https://flowhub.org/f/bd2901f55cfc55ef) demonstrates the use of the [GitHub functionality](https://flowhub.org/f/d0506e991d512ace) and also the [Restart](https://flowhub.org/f/4ffc58e357efa2a0) flow - both are referenced via link out nodes from within the flow. (this [link out node](https://cdn.flowhub.org/?t=0&fhid=bd2901f55cfc55ef&tk=#node/f5e06b7b8845273b) for the github functionality and this [link out node](https://cdn.flowhub.org/?t=0&fhid=bd2901f55cfc55ef&tk=#node/bd5d75b0a94c5d00) for restart).

To include those flows, there are two FlowHub pull within the [flow](https://flowhub.org/f/bd2901f55cfc55ef) nodes: [this one for Github functionality](https://cdn.flowhub.org/?t=0&fhid=bd2901f55cfc55ef&tk=#node/32aba161ccb9b03c) and this [one for the restart](https://cdn.flowhub.org/?t=0&fhid=bd2901f55cfc55ef&tk=#node/c13227aa7da797b0). FlowHub pull nodes are designed to be [opened and the there is a pull button](https://cdn.flowhub.org/?t=0&fhid=bd2901f55cfc55ef&tk=#node/32aba161ccb9b03c/edit) to retrieve and include the relevant flow. I.e., they define flow dependencies amongst flows.

**Comment by gorenje** [comment]
> And yes the documentation around GITHUB_TOKEN is admittedly thin ...

What I'll do is add a field to the NodeDevOps node so that a location of the GITHUB_TOKEN can be choosen. As is the case with the OTP generator sidebar:

![Screenshot 2024-10-16 at 09 37 39](https://github.com/user-attachments/assets/b17475e0-ea9f-4ec9-afa1-627b362ab9a0)

I'll also do that with the npm token that is required for publishing.

Update:

This is now fixed with [0.4.7](https://flows.nodered.org/node/@gregoriusrippenstein/node-red-contrib-nodedev) which allows for the specification of github and npm auth tokens in the dialog for the [NodeDevOps node](https://github.com/gorenje/node-red-contrib-nodedev/blob/2c7b4dfd156e5196e6c0ac48e08e649e1a22ae0b/nodes/05-node-dev-ops.html#L313-L319):

![Screenshot 2024-10-16 at 12 41 04](https://github.com/user-attachments/assets/d3d9116d-f175-4f7c-b03d-1a9d910ba7e6)

**Comment by bohtho** [comment]
It's not that ENV variables were a surprise, but your latest iteration is much less convoluted for new users, and this node's offer of a more "visual coding experience" for nodes wouldn't attract the most hardcore commandline  hackers either. Nice!👍🏻 

While on this (vague) subject of "missing nodes"; Are you aware of any documented attempts to elegantly mix the nodedev and nodemaker nodes in a single node generating flow? I think it's a good idea for a visual node editor to offer some help with building the ui too, but in this case they both have helpful strong points and don't necessarily overlap as much as it may seem from the names.

**Comment by gorenje** [comment]
I tried out nodemaker a [couple](https://flowhub.org/f/e4ec4a4c862fd1a8) of [times](https://flowhub.org/f/2c6bf243745407b7). I only used it to do an initial UI and then integrated my results into what I was building. I don't know whether [@Steveorevo](https://github.com/Steveorevo) is still maintaining nodemaker, I did notice that there have been changes to the NR components making nodemaker a little inconsistent with the current versions of NR. But it still works out of the box, definitely good stuff 👍 

As for NodeDev, it's definitely hard work to find a "simple" method for supporting node development and perhaps it's not even needed. I only started work on NodeDev because I got tired of the context switch between my text editor and my visual editor. And the constant need to reinstall packages into Node-RED on the commandline and then having to restart NR. In addition, I found the npm publish flow to be a pain on the commandline (OTP generations and auth tokens ...).

I just wanted to automate all the fluff around NR node development and hence NodeDev was created. Then I had to [bootstrap](https://en.wikipedia.org/wiki/Bootstrapping) it because I couldn't have it depend on itself and also I didn't want to develop NodeDev using my text editor. So that got complicated, i.e. the [flow](https://flowhub.org/f/b92be5062203ff69) but the end result is simpler - for me - node development.

Thanks for taking the effort to have a look at NodeDev 👍
